doc: renamed doc files to .md, adapted build scripts

eljefedelrodeodeljefe opened the pull request:
This is for consistency and a little less noise for the eye
as in vm.markdown (2 against 8 chars)

Discussion on: nodejs/node#4726 and nodejs/node#4733

PR-URL:
Reviewed-By:
Reviewed-By:

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/1298/

cjihrig commented:
Is this what @rvagg was referring to in #4733 when he said:

> Regarding renaming, I don't think you're going to get any support on this, it's unnecessary churn that makes the history too hard to follow. In general churn without strong justification gets a -1 around here for this reason.

misterdjules commented on doc/api/console.md:
```diff
@@ -7,7 +7,7 @@ JavaScript console mechanism provided by web browsers.
 The module exports two specific components:
 
-* A `Console` class with methods such as `console.log()`, `console.error()` and
+* A `ConsoleJefe` class with methods such as `console.log()`, `console.error()` and
```
Typo, this should still be `Console`.

misterdjules replied:
The rest of the changes in this file should also be removed, as this PR is supposed to be about renaming files only.

jasnell replied:
Yeah, it looks like this was done as a branch off the other PR branch so a lot of those other changes were carried over.

eljefedelrodeodeljefe replied:
Yeah. Give me a second to chop the PRs. Git is funny about certain things.

jasnell commented:
@eljefedelrodeodeljefe ... I appreciate the new PR to isolate this one change but this will need to be rebased off master and cleaned up to remove the additional changes from the other PR before it could land at all. Also, as @cjihrig and @rvagg indicate, I'm not sure if there's going to be enough consensus to get this landed. Personally I'm fine with renaming these, however.

eljefedelrodeodeljefe commented:
@jasnell after the comments above I figured that... will likely close this in favor for a new one in a bit. Sorry for the noise. 

About consensus: I know. It just seemed logic to me, while doing the other PR. I am okay with it not landing.

jasnell commented:
No worries. Thank you for bearing with us @eljefedelrodeodeljefe ! The contributions are very welcome, it's just that sometimes it takes some tweaking here and there to get it just right! :-)

rvagg commented:
thanks @eljefedelrodeodeljefe, having a place to discuss this separately is a good idea

eljefedelrodeodeljefe commented:
@rvagg @jasnell that's alright for me. Just one thing: I'd like to have this closed or "paused" until #4733 is closed or merged. Because then I could just `cherry-pick` this commit. It's a lot easier for me of course :) I would have the same procedure for the typo fixes on #4733. When I would need to check out from master, and then cherry-pick from the PR there would be tons of merge conflicts.

jasnell commented:
@eljefedelrodeodeljefe ... understood. I've applied the `in progress` label which should notify folks that it's a work in progress.

eljefedelrodeodeljefe commented:
Thx @jasnell. Just a heads up. I am running:
```
    git filter-branch --tree-filter '  
    if [ -f doc/api/_toc.markdown ]  
    then
       mv doc/api/_toc.markdown doc/api/_toc.md
    fi' --force HEAD  
```
Thi preserves the log, but changed filenames for all commits that  refer to it. With `node` that's an operation that runs on 12k+ files. So for all the files it's probably gonna take another 20hrs on my machine. The history though looks good.

![e0113773eeb4cca772366ed111558d2b08cb93ed_diff_ __users_jefe_documents_repos_node_upstream](https://cloud.githubusercontent.com/assets/3899684/12507343/7b899374-c0f5-11e5-97be-97d7fa21e8b2.png)

jasnell commented:
that kind of forced push likely wouldn't work for master.

TheAlphaNerd commented:
@eljefedelrodeodeljefe is there a reason for changing the file names in the git history and not just a single commit?

eljefedelrodeodeljefe commented:
@jasnell Wouldn't the force push only apply to the PR branch and you would see problematic things in the diff? @TheAlphaNerd this would be the consequence of the above command in order to preserve logs, since `git mv` would render the files `git log` with just one commit. You would need to `git log --follow` to see more though. I have seen several threads on the internets discussing this and that Linus is actively refusing to have a decent rename feature (for okay reasons).

eljefedelrodeodeljefe commented:
@jasnell we have talked this PR through in the doc-WG meeting. @TheAlphaNerd pointed out that as soon has you have cleared your backlog for LTS we could have a shot at merging this. Would you give me a ping as soon as this is done?

TheAlphaNerd commented:
Just tried landing this against staging and found the following conflicts

```
Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add <file>..." to mark resolution)

	both modified:   doc/api/tls.md
	both modified:   tools/doc/addon-verify.js
	both modified:   tools/doc/html.js
```

The conflicts themselves are pretty minimal. LGTM for lts

eljefedelrodeodeljefe commented:
I am rebasing the right now.

eljefedelrodeodeljefe commented:
@TheAlphaNerd can you try again?

eljefedelrodeodeljefe commented:
/cc @nodejs/documentation due to active PRs and lots of movements in the docs, quick response would be super appreciated.

TheAlphaNerd commented:
@eljefedelrodeodeljefe github is reporting there are still conflicts to resolve between this and master... have you taken a look at that?

eljefedelrodeodeljefe commented:
rebase now green, `make doc` seems to render fine.

TheAlphaNerd commented:
The only conflict now is in the `Makefile` which is to be expected since there are multiple changes in the make file not present in v4.x

LGTM

techjeffharris commented:
LGTM.  I personally prefer `.md` to `.markdown`, but if CTC decides that there isn't enough substance beyond renames to warrant the changes to the git history, I'm going to have to trust their judgement.

TheAlphaNerd commented:
I'm actually retroactively going to -0 on this... leaning towards -1

Even thought it was discussed above I didn't really think about the effect of completely wiping the history on the docs... although I'm not convinced that with LTS and stable being fairly in tune that is any less of an aesthetic choice than file name length. It feels like the data could turn out to be useful.

eljefedelrodeodeljefe commented:
we'd still have the rev parsing, but what reassured me was @chrisdickinson that there was lots of precedences for removing and renaming before. See 2014. I don't mind too much, would be a nice fresh start though. If CTC decided against it, then we obviously just let it go.

rvagg commented:
Is the Docs WG making a case for this? If so, perhaps we should hear a position from them?

The .markdown personally bothers me, I prefer .md, but it's pretty trivial and the desire for less churn easily overrides that—`git blame` gets very tedious trying to track down when/why something was changed when there's a lot of things moving around inside files and a change in filename adds an extra wrinkle to that process. I think we've set a fairly clear precedent in rejecting or pushing back on changes that introduce churn with little added value.

However, perhaps the Docs WG has a good case to make on why this is a good thing? I'm of the mind that handing more responsibility for docs/api to the Docs WG is a good thing in general and would prefer not to erect barriers to progress being made by people who are passionate about such things. I know others have a bit more of a mind for stricter quality-control but this is all about give and take.

chrisdickinson commented:
This seems like a case where the potential gain is relatively minimal, but the potential loss is also relatively minimal — the gain being that we get to use the more idiomatic extension, `.md`, on both the new and old docs (and, maybe, that new docs don't need to use `.markdown`.) The loss is that it necessitates an extra argument to `git log` to track changes, and necessitates that folks use `git blame $REV` — a mild, but surmountable negative. Notably, no history is lost with this transformation.

It's a tough call, since it's a subtle upside vs. a subtle downside, but I lean towards merging this & backporting — I am not aware of anyone depending on the `.markdown` extension, as most folks seem to be using the generated `.html` or `.json` docs. I tend to think that this is enough of a nit that it will keep coming up, too, so by fixing this now we bolster the upside by saving ourselves time in the future.

stevemao commented:
I agree with @chrisdickinson , 

> The loss is that it necessitates an extra argument to git log to track changes, and necessitates that folks use git blame $REV

I believe it's just a bit hard for new git users. Most people who contribute to this repo are quite advanced. And what are the chances for new users to track the history of the `.md` files.

Although this is a nit PR, I believe the upside > downside.

jasnell commented:
hmm... think I'm tending towards +1 too for the same reasons @chrisdickinson outlines.

bnoordhuis commented:
As a small data point, I still get annoyed almost daily at the cut off history from when we split test/simple into test/parallel and test/sequential.

jasnell commented:
@nodejs/ctc ... what do we want to do with this one?

eljefedelrodeodeljefe commented:
@bnoordhuis I have this with `src/node.js` lately, too :P 
I'll rebase this tomorrow. Maybe after your next meeting we have a decision.

chrisdickinson commented:
I lean towards renaming. There are workarounds for viewing complete history, & as we come up on v6 we have a window to get this done before the next LTS candidate is cut.

indutny commented:
I guess `git blame` won't work with this? Sorry, but -1

chrisdickinson commented:
@indutny `git blame` takes a revision at which to blame.

indutny commented:
@chrisdickinson which means that I will have to do two actions instead of one.

evanlucas commented:
I don't see it as a justified change. I think I'm -1 on this too.

chrisdickinson commented:
@indutny @evanlucas This issue comes up repeatedly, and fixing it is less invasive than similar changes we've done to `src/node.js` and `test`. Aside from LTS concerns (which we can get out ahead of this is changed before the v6 release), are we `git blame`-ing the docs that often?

evanlucas commented:
@chrisdickinson That's fair. tbh, up until more recently, I was not. I have git blamed the docs on quite a few occasions over the past few weeks though. If this is something we end up doing, I agree it should be done before v6 goes out though...

indutny commented:
@chrisdickinson not more often than the other parts of the code. Sometimes it is the only way to figure out when the feature was deployed. Anyway, my point is that this PR makes one thing significantly less usable, while making the other non-critical thing a bit more consistent.

jbergstroem commented:
But doesn't git blame follow redirects pretty well? Unless I'm missing something I'm `+1` to moving to `.md` because it's "the right thing".

(looking at https://github.com/nodejs/node/commit/0634311a9ac3dda0e3b494f7f510c23ca48acd46)
```bash
nodejs/src ±master $ git log -n 1 --pretty="format:%h (%an <%ae>): %s" -- test/parallel/test-cluster-disconnect-suicide-race.js
0634311 (Rich Trott <rtrott@gmail.com>): test: move cluster tests to parallel

nodejs/src  ±master $ git blame -L1,+5 nodejs/master test/parallel/test-cluster-disconnect-suicide-race.js
9571be12 test/sequential/test-cluster-disconnect-suicide-race.js (Santiago Gimeno 2015-12-18 19:13:50 +0100 1) 'use strict';
44aba1ab test/sequential/test-cluster-disconnect-suicide-race.js (Rich Trott      2016-01-17 23:09:26 -0800 2)
44aba1ab test/sequential/test-cluster-disconnect-suicide-race.js (Rich Trott      2016-01-17 23:09:26 -0800 3) // Test should fail in Node.js 5.4.1 and pass in later versions.
44aba1ab test/sequential/test-cluster-disconnect-suicide-race.js (Rich Trott      2016-01-17 23:09:26 -0800 4)
9571be12 test/sequential/test-cluster-disconnect-suicide-race.js (Santiago Gimeno 2015-12-18 19:13:50 +0100 5) const common = require('../common');
```

indutny commented:
@jbergstroem I think you are right, at least man page agrees with you! :) +1 then, disregard my earlier -1

TheAlphaNerd commented:
I'm +1 on this as well if everyone is on board

eljefedelrodeodeljefe commented:
Rebased. Since there was a change in `node.cc` which introduced a comment hinting to a .markdown page (changed that) this would require CI.

Also I noticed that `make doc` throws errors in js during json generation. They are unrelated to this commit and have been introduced somewhen before. HTML looks fine.

jasnell commented:
I'm +1 as well.

stevemao commented:
Can we get more +1s and get this landed?

jbergstroem commented:
Just to be clear: LGTM

jasnell commented:
LGTM. needs a rebase

stevemao commented on doc/api/index.markdown:
```diff
@@ -1 +0,0 @@
-@include _toc.markdown
```
This should be renamed instead of deletion and addition?

eljefedelrodeodeljefe commented:
Rebased this. HTML looks good. @stevemao I have changed this to several commits, however git isn't clever enough when renaming one-line files. The PR diffs still shows this as deletion/creation, single commits don't though. For better history all three commits should be landed then.

addaleax commented:
LGTM

eljefedelrodeodeljefe commented:
Already needs rebase again. If someone would decide and volunteer to merge hit me up. I'll rebase quickly before that then.

TheAlphaNerd commented:
@eljefedelrodeodeljefe if you rebase this and push I will follow up with the merge right after

eljefedelrodeodeljefe commented:
@TheAlphaNerd done and thanks!

jasnell commented:
So right now would be the perfect time to land a few more doc PRs right? ;-) ... I guess I can wait until after this lands lol

TheAlphaNerd commented:
Just landed locally... doing some testing and will push in a minute

TheAlphaNerd commented:
@eljefedelrodeodeljefe before I push have you had a chance to respond to https://github.com/nodejs/node/pull/4747#discussion_r60178853

eljefedelrodeodeljefe commented:
@TheAlphaNerd was commenting here https://github.com/nodejs/node/pull/4747#issuecomment-212355143. Git isn't clever enough. That's why I did it in 4 commits. When you don't squash everything should be fine. Otherwise we would need 2 PRs

TheAlphaNerd commented:
@eljefedelrodeodeljefe thanks for the heads up :smile: ... incoming

TheAlphaNerd commented:
landed in cff2a13...e546beb

eljefedelrodeodeljefe commented:
Awesome, thx everyone. Even though it's churn here it stands for renovation, yay!

![churn](https://cloud.githubusercontent.com/assets/3899684/14693157/c4e3021e-075c-11e6-9e48-6ee829d3a63a.png)

TheAlphaNerd commented:
@eljefedelrodeodeljefe how long have you been sitting on that graphic?

eljefedelrodeodeljefe commented:
I had to download and  install Illustrator for an hour or two 😤

stevemao commented:
Sorry for the late response @eljefedelrodeodeljefe , even [git-mv]( https://git-scm.com/docs/git-mv ) didn't work?

eljefedelrodeodeljefe commented:
@stevemao it is `git mv`. I think it's just `git`'s or GitHubs diffing that shows it wrong.

silverwind commented:
@eljefedelrodeodeljefe can you open a backport PR for this against v4.x? Would make landing doc stuff on LTS easier.

TheAlphaNerd commented:
@eljefedelrodeodeljefe ping re backport :smile:

eljefedelrodeodeljefe commented:
Sorry, yes, just haven't had a chance to look into the procedure. Will likely do it tomorrow. Sorry for the delay.

eljefedelrodeodeljefe commented:
@TheAlphaNerd #7507 cc @silverwind
